integration/count_compat_test.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package integration
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+
+	"github.com/FerretDB/FerretDB/integration/setup"
+	"github.com/FerretDB/FerretDB/integration/shareddata"
+)
+
+// countCompatTestCase describes count compatibility test case.
+type countCompatTestCase struct {
+	filter     bson.D                   // required
+	resultType compatTestCaseResultType // defaults to nonEmptyResult
+}
+
+// testCountCompat tests count compatibility test cases.
+func testCountCompat(t *testing.T, testCases map[string]countCompatTestCase) {
+	t.Helper()
+
+	// Use shared setup because find queries can't modify data.
+	// TODO Use read-only user. https://github.com/FerretDB/FerretDB/issues/1025
+	s := setup.SetupCompatWithOpts(t, &setup.SetupCompatOpts{
+		Providers:               shareddata.AllProviders(),
+		AddNonExistedCollection: true,
+	})
+	ctx, targetCollections, compatCollections := s.Ctx, s.TargetCollections, s.CompatCollections
+
+	for name, tc := range testCases {
+		name, tc := name, tc
+		t.Run(name, func(t *testing.T) {
+			t.Helper()
+
+			t.Parallel()
+
+			filter := tc.filter
+			require.NotNil(t, filter, "filter should be set")
+
+			var nonEmptyResults bool
+			for i := range targetCollections {
+				targetCollection := targetCollections[i]
+				compatCollection := compatCollections[i]
+				t.Run(targetCollection.Name(), func(t *testing.T) {
+					t.Helper()
+
+					var targetRes, compatRes bson.D
+					targetErr := targetCollection.Database().RunCommand(ctx, bson.D{
+						{"count", targetCollection.Name()},
+						{"query", filter},
+					}).Decode(&targetRes)
+					compatErr := compatCollection.Database().RunCommand(ctx, bson.D{
+						{"count", compatCollection.Name()},
```

Comment:
Why not CountDocuments?

Reply:
When I did something like that somewhere else, you asked me, "Why not `RunCommand`?" :) 

My answer is that `RunCommand` is more universal, and it's easy to extend it if needed. But I don't mind using `CountDocuments` if there is a reason why it's better.

Reply:
By the same logic we should use RunCommand with "find" instead of Find and FindOne methods. This way, we may ensure that out implementation matches our tests… and maybe does not actually driver's Find and FindOne implementation. But we care about real applications first, and they use drivers, so we should test with drivers where possible, not with raw commands.

Reply:
> When I did something like that somewhere else, you asked me, "Why not RunCommand?" :)

IIRC, that was a question for command without driver's method

Reply:
@AlekSi func fact: `CountDocuments` wraps aggregation and calls aggregation, not the `count` command. So we can't use it in this test for the `count` command.

Reply:
Great, please add a comment about that in the next PR

Reply:
added todo

Reply:
fixed